src/audio/rtnr/rtnr.c

```diff
@@ -768,97 +735,94 @@ static int rtnr_copy(struct comp_dev *dev)
 
 	comp_dbg(dev, "rtnr_copy()");
 
-	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
-
 	/* put empty data into output queue*/
-	RTKMA_API_First_Copy(cd->rtk_agl, cd->source_rate,
-			     audio_stream_get_channels(&source->stream));
+	RTKMA_API_First_Copy(cd->rtk_agl, cd->source_rate, audio_stream_get_channels(source));
 
-	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	if (!frames)
+		return 0;
 
+	comp_dbg(dev, "rtnr_copy() frames = %d", frames);
 	if (cd->process_enable) {
-		frames = audio_stream_avail_frames(&source->stream, &sink->stream);
-
-		/* Process integer multiple of RTNR internal block length */
-		frames = frames & ~RTNR_BLK_LENGTH_MASK;
-
-		comp_dbg(dev, "rtnr_copy() source_id: %d, frames = %d", buf_get_id(source),
-			 frames);
+		/* Run processing function */
 
-		if (frames) {
-			source_bytes = frames * audio_stream_frame_bytes(&source->stream);
-			sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);
+		/* copy required data from sof audio stream to RTNR audio stream */
+		rtnr_copy_from_sof_stream(sources_stream[0], source);
+		rtnr_copy_from_sof_stream(sink_stream, sink);
```

Comment:
is this correct? Need to copy data from a sink stream before processing?

Reply:
The copy function (I've not changed functions names) is converting between SOF stream format and Realtek's library stream format. There's similarity but they are not exactly same. This only converts struct members. It's not moving and PCM data.

Reply:
@singalsu got it, thanks!